File: 백준/Gold/14890. 경사로/경사로.py
import sys
input = sys.stdin.readline

n, l = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]

def check(arr):
    visited = [False for _ in range(n)]
    for i in range(n-1):

        # 현재칸, 다음칸의 높이차가 1 초과일 시 False 반환
        if abs(arr[i] - arr[i+1]) > 1:
            return False
        
        # 현재칸, 다음칸의 높이가 같으면 일단 보류
        if arr[i] == arr[i+1]:
            continue
        
        # 현재칸의 높이가 다음칸의 높이보다 클 때
        elif arr[i] > arr[i+1]:

            # 경사로의 길이만큼 다음칸을 검사할 것
            for j in range(i+1,i+l+1):

                # 다음 칸들이 0~n-1의 index를 가질 때만 검사해야한다.
                if 0 <= j < n:

                    # 경사로를 설치할 칸의 높이가 다르면 False 반환
                    if arr[j] != arr[i+1]:
                        return False
                    
                    # 이미 경사로를 설치한 칸이라면 False 반환
                    if visited[j] == True:
                        return False
                    
                    # 조건 만족 시 경사로 설치
                    visited[j] = True
                
                # index가 범위를 벗어날 경우
                else:
                    return False
        # 현재칸의 높이가 다음칸의 높이보다 작을 때
        elif arr[i] < arr[i+1]:

            # i부터 거꾸로 l길이 만큼의 경사로를 설치해야한다.
            for j in range(i, i-l, -1):
                if 0 <= j < n:
                    # 경사로를 설치할 칸의 높이가 다르면 False 반환
                    if arr[j] != arr[i]:
                        return False
                    
                    if visited[j] == True:
                        return False
                    
                    visited[j] = True

                else:
                    return False
    return True
          
cnt = 0

# 행 검사
for i in range(n):
    arr = graph[i]
    if check(arr) == True:
        cnt += 1

# 열 검사
for j in range(n):
    arr = []
    for i in range(n):
        arr.append(graph[i][j])
    if check(arr) == True:
        cnt += 1

print(cnt)

File: 백준/Gold/14890. 경사로/경사로.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int l;
    static int[][] graph;
    static int cnt;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());

        graph = new int[n][n];
        cnt = 0;

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        // 행 검사
        for (int i = 0; i < n; i++) {
            int[] arr = graph[i];
            if (check(arr)) {
                cnt++;
            }
        }

        // 열 검사
        for (int j = 0; j < n; j++) {
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = graph[i][j];
            }
            if (check(arr)) {
                cnt++;
            }
        }

        System.out.println(cnt);
    }

    static boolean check(int[] arr) {
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n - 1; i++) {
            if (Math.abs(arr[i] - arr[i + 1]) > 1) {
                return false;
            } else if (arr[i] == arr[i + 1]) {
                continue;
            } else if (arr[i] > arr[i + 1]) {
                for (int j = i + 1; j <= i + l; j++) {
                    if (j >= n || arr[j] != arr[i + 1] || visited[j]) {
                        return false;
                    }
                    visited[j] = true;
                }
            } else if (arr[i] < arr[i + 1]) {
                for (int j = i; j > i - l; j--) {
                    if (j < 0 || arr[j] != arr[i] || visited[j]) {
                        return false;
                    }
                    visited[j] = true;
                }
            }
        }
        return true;
    }
}